cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -723,17 +679,100 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
-	if ip, ok := p.(*process.Init); ok {
-		// Ensure all children are killed
-		if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
-			if err := ip.KillAll(s.context); err != nil {
-				log.G(s.context).WithError(err).WithField("id", ip.ID()).
-					Error("failed to kill init's children")
+// handleInitExit processes container init process exits.
+// This is handled separately from non-init exits, because there
+// are some extra invariants we want to ensure in this case, namely:
+// - for a given container, the init process exit MUST be the last exit published
+// This is achieved by:
+// - killing all running container processes
+// - waiting for the container's running exec counter to reach 0, or timeout
+// - removing all remaining processes from s.running so that we s.processExits
+// doesn't publish their exits
+// - publishing the received exec exits, and then publishing the init exit.
+func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
+	// kill all running container processes
+	if err := p.KillAll(s.context); err != nil {
+		log.G(s.context).WithError(err).WithField("id", p.ID()).
+			Error("failed to kill init's children")
+	}
+
+	s.lifecycleMu.Lock()
+	defer s.lifecycleMu.Unlock()
+	numRunningExecs := s.runningExecs[c]
+	if numRunningExecs == 0 {
+		s.handleProcessExit(e, c, p)
+		return
+	}
+
+	events := make(chan int, numRunningExecs)
+	s.execCountSubscribers[c] = events
+
+	go func() {
+		defer func() {
+			s.lifecycleMu.Lock()
+			defer s.lifecycleMu.Unlock()
+			delete(s.execCountSubscribers, c)
+		}()
+
+		// wait for running processes to exit
```

Comment:
Two things:

1) This is just exec processes, not init's children or exec's children, right?  The shim doesn't know about child processes.
2) I'm not sure the timeout approach is sound.  Processes can become unkillable (both in a private pidns and shared pidns) via TASK_UNINTERRUPTIBLE (which typically seems to be for uninterruptible waits on I/O, but I'm not sure if other cases exist).  If we do this timeout we can get into a situation where (a) the shim believes the container is no longer running (i.e., has published the init exit and no longer processes additional exits), (b) processes still exist inside resources related to the container (pidns, cgroup, etc), (c) those resources may not be possible to clean up, (d) ...then what?  I'm worried that this creates another unrecoverable state even if the stuck processes recover from TASK_UNINTERRUPTIBLE.

Reply:
1. yes, the shim doesn't know about the others.
2. The timeout approach is more of a holdout in case something goes wrong – to ensure that regardless, IF we've received an init exit, we don't delay processing it too much/eventually we do guarantee that it's published. I'm not sure what's more correct in this case – publish it even if there's still a running process, or never publish it at all even though we it was reaped. Regardless, I (think) this should only apply in the shared-namespace case as if the init process has been reaped, all other processes would have been reaped first in the private namespace case

Reply:
Another thought – would the timeout be more palatable if we increased the watchdog timeout to longer intervals – such as 30 seconds?

Reply:
> Another thought – would the timeout be more palatable if we increased the watchdog timeout to longer intervals – such as 30 seconds?

The kernel doesn't provide a guarantee that a TASK_UNINTERRUPTIBLE process will be able to resume within a given timeframe (or ever); I have a bug report showing processes in this state for over 2 days before the VM was terminated.

I think the problem is that we're conflating two different issues:
1. Exits can be lost when there are execs that race with the init exit
2. Some processes may not be terminated quickly (in either private-pidns or shared-pidns scenarios)

My preference would be that we restrict the fix here to just solving the known race condition (1).  If that means delaying the exit publishing indefinitely if there are still unterminated execs...I think that's reasonable so long as the exit never actually gets _lost_.

cc @dmcgowan in case you have other thoughts here

Reply:
Just FYI:
> Some processes may not be terminated quickly (in either private-pidns or shared-pidns scenarios)

In the private-pidns scenario, the [kernel assures us](https://elixir.bootlin.com/linux/v6.7.4/source/kernel/pid_namespace.c#L237) that by the time we reap the init process of the namespace, all other processes are already gone. This means that we'd never run into this scenario (waiting for other processes to exit) in the private pidns case.

In any case, I'm totally okay with us just waiting in this case.

Reply:
> In the private-pidns scenario, the kernel assures us that by the time we reap the init process of the namespace, all other processes are already gone. This means that we'd never run into this scenario (waiting for other processes to exit) in the private pidns case.

I was convinced that this wasn't the case...and then I looked again at the sosreport I have from a VM with TASK_UNINTERRUPTIBLE and I do still see that pidns's pid 1 in `/proc` even though it has disappeared from `cgroup.procs`.